### PR TITLE
changes in trigger api to include trigger for `cloud-pat-individual` configuration flow

### PR DESCRIPTION
**Pull request by MuskanPaliwal**
## Summary by CodeRabbit

- **New Features**
  - Enhanced the build setup process to include selections for provider, installation type, and hosting, improving customization for user projects.
  - Integrated encryption and decryption for sensitive data to enhance security.
  - Added functionality to trigger Cloud Build using Google Cloud API with encryption for GitHub personal access token.

### CODE REVIEW

File: utils/encrypt_decrypt.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+export const generateSecretKey = (): string => {
+    return crypto.randomBytes(32).toString('hex'); // 256-bit key
+}
+
+export const encrypt = (secretKey: string, data: string): string => {
+    try {
+        // Generate an initialization vector (IV) from the secret key
+        const iv = crypto.randomBytes(16);
+        // Create an encryption cipher
+        const cipher = crypto.createCipheriv('aes-256-cbc', Buffer.from(secretKey, 'hex'), iv);
```
[the remark this answers is not in the document]
Comment (tapishr) [reply utils/encrypt_decrypt.ts:12]: Sounds good

File: utils/pubsub/pubsubClient.ts [hunk 2]
```diff
@@ -172,4 +173,67 @@ export async function pollBuildStatus(projectId: string, location: string, build
...
+        }
+    }, {
+        headers: {
+            'Authorization': `Bearer ${accessToken}`,
+            'Content-Type': 'application/json'
+        },
+    });
+
+    if (!response.data) {
+        console.error('[triggerBuildUsingGcloudApi] Error triggering build: ', response.statusText);
```
Comment (tapishr) [inline utils/pubsub/pubsubClient.ts:231]: function tag fix

File: pages/api/dpu/trigger.ts [hunk 3]
```diff
@@ -50,12 +57,21 @@ const triggerHandler = async (req: NextApiRequest, res: NextApiResponse) => {
...
-	console.info("[triggerHandler] build status: ", buildStatus);
+	let buildStatus: CloudBuildStatus;
+	if (jsonBody.selectedProvider === 'github' && jsonBody.selectedInstallationType === 'individual' && jsonBody.selectedHosting === 'cloud'){
+		buildStatus = await triggerCloudPatBuildUsingGcloudApi(jsonBody.userId, topicName, jsonBody.github_pat, jsonBody.selectedProvider).catch(err => {
+			console.error(`[triggerHandler] error in triggering build`, err);
+			return { success: false, message: 'Unable to trigger build using GCloud API' };
+		});
+		console.info("[triggerHandler] build status: ", buildStatus);
+	
+	} else {
```
Comment (tapishr) [inline pages/api/dpu/trigger.ts:68]: else if  - github,project,cloud or bitbucket,project,cloud

File: utils/pubsub/pubsubClient.ts [hunk 4]
```diff
@@ -172,4 +173,67 @@ export async function pollBuildStatus(projectId: string, location: string, build
     }
 	console.error('[pollBuildStatus] Build status check exceeded maximum attempts');
     return 'TIMEOUT'; // Indicate a timeout in build status check
+}
+
+export async function triggerCloudPatBuildUsingGcloudApi(user_id: string, topic_name: string, encrypted_github_pat: string, provider: string): Promise<CloudBuildStatus> {
```
Comment (tapishr) [inline utils/pubsub/pubsubClient.ts:178]: This and the other function can be written as one function with the number and type of variables different for the two cases

File: utils/trigger.ts [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        const buildDetails = response.data.metadata.build;
+        console.info(`[triggerCommonCloudBuild] Build triggered successfully for topic_name: ${topic_name} and user_id: ${user_id} with build details: ${buildDetails.id}`);
+        return { success: true, message: 'Build triggered successfully', buildDetails };
+    }).catch(error => {
+        console.error('[triggerCommonCloudBuild] Error in triggering Cloud Build:', error.message);
+        return { success: false, message: `Error in triggering Cloud Build: ${error.message}`, buildDetails: null };
+    });
+}
+
+export async function triggerBuildUsingGcloudApi(user_id: string, topic_name: string): Promise<CloudBuildStatus> {
```
Comment (MuskanPaliwal) [inline utils/trigger.ts:137]: better function name

**Comment by MuskanPaliwal** [comment]
- Add  `_SECRET_KEY` var on prod vibinex-server trigger before merging the pr. You will have to generate the key.
-  Add  `_CLOUD_BUILD_PAT_TRIGGER_ID`  on prod vibinex-server trigger. The trigger name is `DPU_USER_PAT`

File: pages/api/dpu/trigger.ts [hunk 6]
```diff
@@ -1,17 +1,25 @@
...
 const triggerHandler = async (req: NextApiRequest, res: NextApiResponse) => {
 	console.info("[triggerHandler] pub sub setup info in db...");
 	const jsonBody = req.body;
 	let topicName: string;
-	if (!jsonBody.userId) {
+	if (!jsonBody.selectedProvider || !jsonBody.selectedInstallationType || !jsonBody.selectedHosting || !jsonBody.userId) {
 		console.error("[triggerHandler] Invalid request body");
-		res.status(400).json({ "error": "Invalid request body" });
-		return;
+		return res.status(400).json({ "error": "Invalid request body" });
```
Comment (tapishr) [inline pages/api/dpu/trigger.ts:13]: Can you separate these lines and maybe return in a new line

File: pages/api/dpu/trigger.ts [hunk 7]
```diff
@@ -50,12 +58,24 @@ const triggerHandler = async (req: NextApiRequest, res: NextApiResponse) => {
...
+	
+	} else if (jsonBody.selectedInstallationType === 'project' && jsonBody.selectedHosting === 'cloud') {
+		buildStatus = await triggerCloudProjectBuildUsingGcloudApi(jsonBody.userId, topicName).catch(err => {
+			console.error(`[triggerHandler] error in triggering build`, err);
+			return { success: false, message: 'Unable to trigger build using GCloud API' };
+		});
+		console.info("[triggerHandler] build status: ", buildStatus);
+	} else {
+		console.error('[triggerHandler] Invalid provider, installation type, or hosting combination');
+		return res.status(400).json({ "error": "Invalid provider, installation type, or hosting combination", success: false });
```
Comment (tapishr) [inline pages/api/dpu/trigger.ts:77]: Same comment for return

**Review by tapishr — DISMISSED** [review]

**Review by tapishr — APPROVED** [review]

**Pull request merged**